tests/Midi_gtest.cpp

```diff
@@ -0,0 +1,288 @@
+    {
+        UNUSED(buff);
+        UNUSED(size);
+    }
+    uint8_t Rx() { return 1; }
+    bool    RxActive() { return true; }
+
+  private:
+};
+
+class MidiTest : public ::testing::Test
+{
+  protected:
+    void SetUp()
+    {
+        MidiHandler<MidiTestTransport>::Config conf;
+        midi.Init(conf);
+    }
+
+    //help with parsing messages
+    void Parse(uint8_t* msgs, int size)
+    {
+        for(int i = 0; i < size; i++)
+        {
+            midi.Parse(msgs[i]);
+        }
+    }
+
+    //help with parsing messages
+    void ParseAndPop(uint8_t* msgs, int size)
+    {
+        Parse(msgs, size);
+        event = midi.PopEvent();
+    }
+
+    //help with parsing sysex messages
+    void ParseAndPopSysex(uint8_t* msgs, int size)
+    {
+        midi.Parse(0xf0); //sysex
+        Parse(msgs, size);
+        midi.Parse(0xf7); //end of sysex
+        event = midi.PopEvent();
+    }
+
+    //help testing
+    void Test(uint8_t evType,
+              uint8_t chkType,
+              uint8_t chn,
+              bool    twoData,
+              uint8_t d0,
+              uint8_t d1)
+    {
+        //test eventtype == expected type
+        EXPECT_EQ(evType, chkType);
+
+        //test channel
+        EXPECT_EQ(event.channel, chn);
+
+        //test data
+        EXPECT_EQ(event.data[0], d0);
+        if(twoData)
+        {
+            EXPECT_EQ(event.data[1], d1);
+        }
+
+        TestEmptyQueue();
+    }
+
+    //help test sysex
+    void TestSysex(uint8_t* msgs, int size)
+    {
+        EXPECT_EQ((uint8_t)event.type, (uint8_t)SystemCommon);
+        EXPECT_EQ((uint8_t)event.sc_type, (uint8_t)SystemExclusive);
+
+        EXPECT_EQ(event.sysex_message_len, size);
+
+        for(int i = 0; i < size; i++)
+        {
+            EXPECT_EQ(event.sysex_data[i], msgs[i]);
+        }
+
+        TestEmptyQueue();
+    }
+
+    //test for empty queue
+    void TestEmptyQueue()
+    {
+        while(midi.HasEvents())
+        {
+            ADD_FAILURE() << "Queue not empty";
+            midi.PopEvent();
+        }
+    }
+
+    MidiHandler<MidiTestTransport> midi;
+    MidiEvent                      event;
+};
+
+//Channel Voice messages
+//for now I'm not using asNoteOff, etc. easier to just test on the raw MidiEvent
+TEST_F(MidiTest, channelVoice)
+{
+    for(uint8_t type = 0; type < 7; type++)
+    {
+        for(uint8_t chn = 0; chn < 16; chn++)
+        {
+            for(uint8_t d0 = 0; d0 < 128; d0++)
+            {
+                for(uint8_t d1 = 0; d1 < 128; d1++)
+                {
+                    uint8_t msgs[]
+                        = {(uint8_t)(0x80 + (type << 4) + chn), d0, d1};
+
+                    bool sendThree = type != 4 && type != 5;
+                    ParseAndPop(msgs, 2 + sendThree);
+
+                    uint8_t chkType
+                        = (type == 3 && d0 > 119) ? (uint8_t)ChannelMode : type;
+                    Test((uint8_t)event.type, chkType, chn, sendThree, d0, d1);
```

Comment:
I would prefer to have the `EXPECT_EQ()` statements directly here in the test body instead of in the `Test()` function.

- makes it easier to understand what the test does when reading the test body. I know it seems logical to extract this into a shared function when it occurs in multiple tests. But as a reader, you'll have to scroll up and trace which variable from the test ends up in which `EXPECT_EQ` to understand what's going on. 
- when the test fails, the output will contain the line number of the `EXPECT_` macro that failed. If this line number is burried in a function somewhere, it's harder to trace back why it failed.